src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,200 @@
-        return null;
+        checkPositionIndex(index, SETTING);
+        Node<T> node = node(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkPositionIndex(index, SETTING);
+        Node<T> removedNode = node(index);
+        if (index == 0) {
+            first = first.next;
+            if (first != null) {
+                first.prev = null;
+            } else {
+                last = null;
+            }
+        } else if (index == size - 1) {
+            last = last.prev;
+            if (last != null) {
+                last.next = null;
+            } else {
+                first = null;
+            }
+        } else {
+            unlink(removedNode);
+        }
+        size--;
+        return removedNode.item;
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> itemNode = first; itemNode != null; itemNode = itemNode.next) {
+            if ((itemNode.item == null && object == null)
+                    || (itemNode.item != null && itemNode.item.equals(object))) {
+                if (itemNode == first) {
+                    first = first.next;
+                    if (first != null) {
+                        first.prev = null;
+                    } else {
+                        last = null;
+                    }
+                } else if (itemNode == last) {
+                    last = last.prev;
+                    if (last != null) {
+                        last.next = null;
+                    } else {
+                        first = null;
+                    }
+                } else {
+                    unlink(itemNode);
+                }
+                size--;
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        if (first == null && last == null) {
+            return true;
+        }
+        return size == 0;
+    }
+
+    private void linkLast(T element) {
+        final Node<T> prev = last;
+        final Node<T> newNode = new Node<>(prev,element,null);
+        last = newNode;
+        if (prev == null) {
+            first = newNode;
+        } else {
+            prev.next = newNode;
+        }
+        size++;
+    }
+
+    private void checkPositionIndex(int index, int expressionToCheck) {
+        if (expressionToCheck == NOT_SETTING) {
+            if (!isPositionIndex(index)) {
+                throw new IndexOutOfBoundsException(createIndexOutOfBoundMessage(index));
+            }
+        } else {
+            if (!isPositionIndexToSet(index)) {
+                throw new IndexOutOfBoundsException(createIndexOutOfBoundMessage(index));
+            }
+        }
+    }
+
+    private Node<T> node(int index) {
```

Comment:
It will be better if you rename this method. Maybe finNode or findNodeByIndex()

Reply:
agree